Make aerosol output frequency variable

WalterKolczynski-NOAA opened the pull request:
# Description
The `AERO_HISTORY.rc` file that controls the GOCART output is turned into a template usint `@[ ]` syntax to allow for variable output times. A new "parsing" script is created for GOCART that handles parsing the `AERO_HISTORY.rc` file through `atparse`. Other manip- ulation of these files is also moved to the new parsing script to be consistent with other components.

A new variable, `FHOUT_AERO` is introduced to control the output frequency of the GOCART output. This is initially set to 3 to match other components (this is a change from the current value of 6). Since some aerosol fields are also included as part of the atmos output, it may be desirable to meld this back into `FHOUT` later. However, the atmos output has two frequencies (`FHOUT_HF` and `FHOUT`), a feature not supported by other components.

The RUN lists for aero fcst, aero analysis, and waves is simplified to avoid confusion, since there is no need for them to only be set if the components are on.

Resolves #2072

# Type of change
- [ ] Bug fix (fixes something broken)
- [x] New feature (adds functionality)
- [ ] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO

# How has this been tested?
- S2SWA forecast-only on Hercules (manually modifying `AERO_FCST_RUN`)

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have documented my code, including function, input, and output descriptions
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] This change is covered by an existing CI test or a new one has been added
- [x] I have made corresponding changes to the system documentation if necessary

WalterKolczynski-NOAA committed:
Make aerosol output frequency variable

The `AERO_HISTORY.rc` file that controls the GOCART output is turned
into a template usint `@[ ]` syntax to allow for variable output
times. A new "parsing" script is created for GOCART that handles
parsing the `AERO_HISTORY.rc` file through `atparse`. Other manip-
ulation of these files is also moved to the new parsing script to be
consistent with other components.

A new variable, `FHOUT_AERO` is introduced to control the output
frequency of the GOCART output. This is initially set to 3 to match
other components (this is a change from the current value of 6).
Since some aerosol fields are also included as part of the atmos
output, it may be desirable to meld this back into `FHOUT` later.
However, the atmos output has two frequencies (`FHOUT_HF` and
`FHOUT`), a feature not supported by other components.

The RUN lists for aero fcst, aero analysis, and waves is simplified
to avoid confusion, since there is no need for them to only be set
if the components are on.

Resolves #2072

WalterKolczynski-NOAA committed:
Correct shellcheck error

DavidHuber-NOAA commented on parm/config/gfs/config.base:
```diff
@@ -176,9 +176,9 @@ export DO_OCN="NO"
-export AERO_FCST_RUN="" # When to run aerosol forecast: gdas, gfs, or both
-export AERO_ANL_RUN="" # When to run aerosol analysis: gdas, gfs, or both
-export WAVE_RUN="" # When to include wave suite: gdas, gfs, or both
+export AERO_FCST_RUN="gdas" # When to run aerosol forecast: gdas, gfs, or both
+export AERO_ANL_RUN="both" # When to run aerosol analysis: gdas, gfs, or both
+export WAVE_RUN="both" # When to include wave suite: gdas, gfs, or both
```
I like this!  So much simpler than hunting down the correct location within `APP =~ ...A` if blocks.

CoryMartin-NOAA replied:
Same!

DavidHuber-NOAA commented on ush/parsing_namelists_GOCART.sh:
```diff
@@ -0,0 +1,52 @@
+    local fhout_aero_padded
+    fhout_aero_padded=$(printf "%02d" "${FHOUT_AERO}")
+    # Only instantaneous AOD is output right now
+    local inst_aod_freq="${fhout_aero_padded}0000"
```
I think I am missing something.  Where does `inst_aod_freq` actually get written to `AERO_HISTORY.rc`?

WalterKolczynski-NOAA replied:
`atparse` does it on L34. This is how the other "parsing" scripts work too.

CoryMartin-NOAA reviewed (dismissed):
I didn't test but this looks good to me

aerorahul reviewed (dismissed):
Looks good.
Question:
Not for this PR to address -- how diffferent is the `parm/ufs/gocart/AERO_HISTORY.rc` file in the global-workflow from the one in ufs-weather-model?

WalterKolczynski-NOAA commented:
> Looks good. Question: Not for this PR to address -- how diffferent is the `parm/ufs/gocart/AERO_HISTORY.rc` file in the global-workflow from the one in ufs-weather-model?

@aerorahul Looks like they have already templated the one frequency we use, but not the others. A couple other small changes (also whitespace changes not shown):
```
diff parm/ufs/gocart/AERO_HISTORY.rc sorc/ufs_model.fd/tests/parm/gocart/AERO_HISTORY.rc.IN -b
384c384
<   inst_aod.frequency:  060000,
---
>   inst_aod.frequency:  @[AOD_FRQ],
390a391
>                        'NIEXTTAU'      , 'NI'       , 'AOD_NI',
405,406c406,407
<   tavg_2d_rad.fields:     'CA.bcEXTTAU' , 'CA.bc' ,
< 			  'CA.ocEXTTAU' , 'CA.oc' ,
---
>   tavg_2d_rad.fields:     'CAEXTTAUCA.bc' , 'CA.bc' ,
>                           'CAEXTTAUCA.oc' , 'CA.oc' ,
```

(This is against our develop, not the changes in this PR.)

TerrenceMcGuinness-NOAA commented:
Hera CI System Wide Failure:

```
Cannot contact Hera-EMC: java.lang.InterruptedException

Run rocotorun.
Gather Rocoto statistics
(C48_ATM_8a18d626 on Hera)
	Total Cycles: 1
	Number Cycles done: 0
	State: RUNNING
Run rocotorun.
/apps/rocoto/1.3.7/bin/rocotorun: fork: retry: Resource temporarily unavailable
/apps/rocoto/1.3.7/bin/rocotorun: fork: retry: Resource temporarily unavailable
/apps/rocoto/1.3.7/bin/rocotorun: fork: retry: Resource temporarily unavailable
/apps/rocoto/1.3.7/bin/rocotorun: fork: retry: Resource temporarily unavailable
/apps/rocoto/1.3.7/bin/rocotorun: fork: Resource temporarily unavailable
```

DavidHuber-NOAA commented:
**UPDATED**

The `gdas.yaml.j2` archiving yaml picks up on `AERO_ANL_RUN` only, so this needs to be amended to also check the `DO_AERO` flag.  Suggest changing
https://github.com/NOAA-EMC/global-workflow/blob/a42c833abe81df112e80da35843d698677d58bd7/parm/archive/gdas.yaml.j2#L70-L72
to
```jinja
            {% if (AERO_ANL_RUN == "gdas" or AERO_ANL_RUN == "both") and DO_AERO %}
        - "{{ COMIN_CHEM_ANALYSIS | relpath(ROTDIR) }}/{{ head }}aerostat"
            {% endif %}
```

Something similar would also need to be done for `master_gfs.yaml.j2`:
https://github.com/NOAA-EMC/global-workflow/blob/a42c833abe81df112e80da35843d698677d58bd7/parm/archive/master_gfs.yaml.j2#L36-L41

WalterKolczynski-NOAA commented:
> The `gdas.yaml.j2` archiving yaml picks up on `AERO_ANL_RUN` only, so this needs to be amended to also check the `APP`. Suggest changing
> 
> https://github.com/NOAA-EMC/global-workflow/blob/a42c833abe81df112e80da35843d698677d58bd7/parm/archive/gdas.yaml.j2#L70-L72
> 
> 
> to
> ```
>             {% if (AERO_ANL_RUN == "gdas" or AERO_ANL_RUN == "both") and APP.endswith("A") %}
>         - "{{ COMIN_CHEM_ANALYSIS | relpath(ROTDIR) }}/{{ head }}aerostat"
>             {% endif %}
> ```
> 
> or more explicitly
> 
> ```
>             {% if (AERO_ANL_RUN == "gdas" or AERO_ANL_RUN == "both") and APP in ["ATMA", "ATMWA", "S2SA", "S2SWA"] %}
>         - "{{ COMIN_CHEM_ANALYSIS | relpath(ROTDIR) }}/{{ head }}aerostat"
>             {% endif %}
> ```
> 
> Something similar would also need to be done for `master_gfs.yaml.j2`:
> 
> https://github.com/NOAA-EMC/global-workflow/blob/a42c833abe81df112e80da35843d698677d58bd7/parm/archive/master_gfs.yaml.j2#L36-L41

Why are these not based on `DO_AERO`? That is what they should be based on.

DavidHuber-NOAA commented:
@WalterKolczynski-NOAA `DO_AERO` is a much better option instead of using `APP`, though `AERO_ANL_RUN` and `AERO_FCST_RUN` still need to be referenced for the `aerostat` and forecast files.

DavidHuber-NOAA commented:
I've updated my suggestion above.

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA `DO_AERO` is a much better option instead of using `APP`, though `AERO_ANL_RUN` and `AERO_FCST_RUN` still need to be referenced for the `aerostat` and forecast files.

Presumably, this is only temporary until #2943 is complete. `DO_AERO` should be the only control needed outside of `config.base`.

CoryMartin-NOAA commented:
@WalterKolczynski-NOAA maybe, there will be a need for a GFS aerosol analysis, but GFS fcst will not have GOCART. I'm not sure if that will be solved by that PR or not.

WalterKolczynski-NOAA committed:
Fix issue with determination of gocart archival

The archive job was using `AERO_ANL_RUN` to determine whether to
archive certain files. With the move to a static declaration of that
instead of only turning it on when aerosols are on, an additional
check is needed.

WalterKolczynski-NOAA committed:
More archive modifications

DavidHuber-NOAA commented:
> Presumably, this is only temporary until #2943 is complete. `DO_AERO` should be the only control needed outside of `config.base`.

I addressed this in #2943, though two additional flags will now be used: `DO_AERO_ANL` and `DO_AERO_FCST`.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh

DavidHuber-NOAA commented:
@WalterKolczynski-NOAA Should this be reviewed again or do you want to wait until #2943 is merged?

WalterKolczynski-NOAA commented:
CI Tests set up to run in /lfs/h2/emc/ptmp/walter.kolczynski/PR/PR_2982/RUNTESTS on WCOSS

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh

DavidHuber-NOAA commented:
@TerrenceMcGuinness-NOAA It looks like this PR finished building on Hera and Hercules, but never started running the tests.  Can you please take a look?

TerrenceMcGuinness-NOAA commented:
The build processes was interrupted by an unknown internal Java error on the controller.  Restarted both CI runs.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh

DavidHuber-NOAA commented:
Looks like `RUNTESTS` needs to be cleaned out.

aerorahul commented:
> Looks like `RUNTESTS` needs to be cleaned out.

Is this an issue to be resolved in the CI?  As in, when the CI is restarted (label reapplied after a previous failure), should the RUNDIRS be cleaned out as part of driver?  @TerrenceMcGuinness-NOAA what do you think?

WalterKolczynski-NOAA commented:
Something with the gefs replay is interacting badly with the list of aerosol outputs. Will need to fix it.

WalterKolczynski-NOAA committed:
Fix GOCART output times

GOCART output does not start until the first output time after FHMIN,
so the list of output times should reflect that to avoid trying to
copy files that don't exist.

WalterKolczynski-NOAA committed:
Correct first output time of GOCART

GOCART needs to use the bigger of FHMIN and FHROT when determining
the first output time.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh

WalterKolczynski-NOAA committed:
Make further corrections to gocart forecast hours

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh

WalterKolczynski-NOAA committed:
Fix GOCART output times

Moves computation of GOCART output times to postdet, since FHROT is
not computed until then.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh

WalterKolczynski-NOAA committed:
Make GOCART output times a function

Makes GOCART output times a function since it is needed in two places.

WalterKolczynski-NOAA committed:
Add mixing defaults entry for aero in GEFS

The defulats file for GEFS was missing an entry for aero to ensure
@field@ get parsed at expt setup time. This was added to gfs as
part of #3053, but gefs was missed.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh

WalterKolczynski-NOAA committed:
Assign gocart fhmax before it is adjusted by IAU

DavidHuber-NOAA commented on parm/archive/master_gfs.yaml.j2:
```diff
@@ -33,7 +33,7 @@ datasets:
 {% endif %}
 
-{% if AERO_FCST_RUN == "gfs" or AERO_FCST_RUN == "both" %}
+{% if DO_AERO and (AERO_ANL_RUN == "gfs" or AERO_ANL_RUN == "both") %}
```
```suggestion
{% if DO_AERO and (AERO_FCST_RUN == "gfs" or AERO_FCST_RUN == "both") %}
```

WalterKolczynski-NOAA committed:
Fix control of gfs aod files

WalterKolczynski-NOAA commented:
CI Tests set up to run in /lfs/h2/emc/ptmp/walter.kolczynski/PR/PR_2982/RUNTESTS on WCOSS

DavidHuber-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -726,15 +711,25 @@ GOCART_postdet() {
 }
 
+GOCART_output_fh() {
+  # This has to be called during postdet after FHROT has been set
```
Do you mean `IAU_FHROT`?  Also seeing `IAU_FHROT` recalculated from `config.base` values seems kludge-y/problematic.  Should we open an issue to try and centralize this?

DavidHuber-NOAA replied:
```suggestion
  # This has to be called during postdet after IAU_FHROT has been set
```

WalterKolczynski-NOAA replied:
Sort of. `fhrot` is what matters. Right now it seems we are just using the `IAU_FHROT` variable for all `fhrot` changes, and then these get converted to `FHROT` in [parsing_model_configure](https://github.com/NOAA-EMC/global-workflow/blob/develop/ush/parsing_model_configure_FV3.sh#L25). Really, we should probably be using `FHROT` here and set it to `IAU_FHROT` (or whatever else) if necessary.

DavidHuber-NOAA replied:
OK, thanks for clarifying.

DavidHuber-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -726,15 +711,25 @@ GOCART_postdet() {
+  # This has to be called during postdet after FHROT has been set
+  local aero_min
+  local gocart_output_fh
+  aero_min=$(( ${IAU_FHROT:-0} > FHMIN ? IAU_FHROT + FHOUT_AERO : FHMIN + FHOUT_AERO ))
```
A comment here would be useful to state why the first forecast hour doesn't have aerosol products and why it depends on IAU.

DavidHuber-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -726,15 +711,25 @@ GOCART_postdet() {
   echo "SUB ${FUNCNAME[0]}: Copying output data for GOCART"
 
   # Copy gocart.inst_aod after the forecast is run (and successfull)
   # TODO: this should be linked but there are issues where gocart crashing if it is linked
```
Not part of this PR, but was this an problem with HDF5?  I'll see if I can recreate it and open an HDF5 issue.

WalterKolczynski-NOAA replied:
IIRC, it is something in MAPL. I feel like we've been told a couple different times it was fixed but it wasn't.

DavidHuber-NOAA commented on ush/forecast_predet.sh:
```diff
@@ -724,6 +724,8 @@ GOCART_predet(){
+  # FHMAX gets modified when IAU is on, so keep origianl value for GOCART output
+  GOCART_MAX=${FHMAX}
+
+  # GOCART output times can't be computed here because they may depend on FHROT
```
```suggestion
  # GOCART output times can't be computed here because they may depend on IAU_FHROT in forecast_postdet.sh.
```

DavidHuber-NOAA reviewed (changes requested):
The content looks good, just some suggestions on comments and future issues.

WalterKolczynski-NOAA committed:
Add comment on gocart min fh

DavidHuber-NOAA committed:
Merge branch 'develop' into feature/gocart_variable_fh
